kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala

```diff
@@ -3,6 +3,59 @@ package me.rexim.morganey.ast
 import me.rexim.morganey.church.ChurchNumberConverter._
 import me.rexim.morganey.church.ChurchPairConverter._
 
+import scala.annotation.tailrec
+
+object LambdaTerm {
+
+  private def nestedApplications(app: LambdaApp): (LambdaTerm, List[LambdaTerm]) = {
+    @tailrec
+    def go(app: LambdaApp, acc: List[LambdaTerm]): (LambdaTerm, List[LambdaTerm]) =
+      app match {
+        case LambdaApp(l: LambdaApp, r) => go(l, r :: acc)
+        case LambdaApp(l, r)            => (l, r :: acc)
+      }
+    go(app, Nil)
```

Comment:
No I did not participate in Competitive Programming. :)
I chose between `go`, `helper` and `loop`.